### ISSUE TITLE
add compile error for assigning to rvalues

### ISSUE THREAD

**Issue by user00e00** [ziglang/zig#2735]
I don't understand why only '[]u8' is accepted here, when I want to modify a segment of the 'outText' array. As long as the type and lengths align, shouldn't all these be accepted? Also I can't determine if 'outText[0..6]' is modified at all, even though that particular line compiles.
```
var outText : [100]u8 = undefined;
var row1 = "1 2 3 |";
var row2 = "4 5 6 |";
const row3 = "7 8 9 |";
outText[0..6] = row1[0..];   // OK: found []u8, but doesn't seem to actually modify anything
outText[13..19] = row2;      // error: expected type '[]u8', found '[7]u8'
outText[26..32] = row3[0..]; // error: expected type '[]u8', found '[]const u8'
outText[39..45] = "a b c |"; // error: expected type '[]u8', found '[7]u8'
```

**Comment by Sahnvour** [comment]
Slicing an array, as in `outText[0..6]` creates a _slice_ pointing to the sequence of elements 0 to 6. As explained in [the documentation](https://ziglang.org/documentation/master/#Slices), slices are a pair of a pointer and a length.

So when you create a slice on your array, and assign it another slice, it does not modify any data. First `outText[0..6]` creates an unnamed temporary slice, to which you are assigning the slice created by `row[0..1]`. This resolves to copying the pointer and length of the slice on the right hand side, which does not modify you array.

**Comment by user00e00** [comment]
> So when you create a slice on your array, and assign it another slice, it does not modify any data. First `outText[0..6]` creates an unnamed temporary slice, to which you are assigning the slice created by `row[0..1]`. This resolves to copying the pointer and length of the slice on the right hand side, which does not modify you array.

You explained it really well. Thank you.

Then I think all these "assignments" should give a different compiler error, like `Assigning to slice with "=" is not supported. Use function "std.....assign....(to,from)" instead`. That should be helpful for anyone used to e.g Matlab.

**Comment by Vexu** [comment]
Fixed by  #4985.

Now gives: `error: invalid left-hand side to assignment`
